torch/csrc/autograd/functions/accumulate_grad.h

```diff
@@ -11,7 +11,83 @@ struct TORCH_API AccumulateGrad : public Node {
 
   variable_list apply(variable_list&& grads) override;
 
+  // Given a variable with its current grad as variable_grad, accumulates
+  // new_grad into variable_grad if in place accumulation is possible.
+  // Otherwise, uses 'update_grad' to update the grad for the variable.
+
+  // variable: the variable whose grad we're accumulating.
+  // variable_grad: the current grad for the variable.
+  // new_grad: new grad we want to acummulate for the variable.
+  // num_expected_refs: the number of refs we expect to hold internally
+  //                    such that it is safe to avoid cloning the grad
+  //                    if use_count() of the grad is less than or equal
+  //                    to this value (in addition to post_hooks).
+  // update_grad: Function that is used to update grad for the variable.
+  //              The argument to the function is a Tensor which
+  //              is used to set a new value for the grad.
+  template <typename T>
+  static void accumulateGradAndCallHooks(
+      const Variable& variable,
+      at::Tensor variable_grad,
```

Comment:
!!! Keep this one a reference! You don't move into this one, it's not taken ownership of in the body here, should be a `const Tensor& variable_grad` !!!

Reply:
I would like to not have to make comments like this in the future :>

Reply:
Looks like const ref doesn't work here (I'll keep it as a non-const ref):

```
caffe2/torch/csrc/autograd/functions/accumulate_grad.h:82:23: error: no viable overloaded '+='
        variable_grad += new_grad;
        ~~~~~~~~~~~~~ ^  ~~~~~~~~
caffe2/aten/src/ATen/TensorOperators.h:23:24: note: candidate function not viable: 'this' argument has type 'const at::Tensor', but method is not marked const
inline Tensor& Tensor::operator+=(const Tensor & other) {
                       ^
caffe2/aten/src/ATen/TensorOperators.h:26:24: note: candidate function not viable: 'this' argument has type 'const at::Tensor', but method is not marked const
inline Tensor& Tensor::operator+=(Scalar other) {
```